calyx-opt/src/passes/metadata_table_gen.rs

```diff
@@ -0,0 +1,110 @@
+use crate::traversal::{Action, Named, VisResult, Visitor};
+use calyx_ir::Id;
+use calyx_utils::WithPos;
+use linked_hash_map::LinkedHashMap;
+use std::fmt;
+use std::path::PathBuf;
+
+/// Metadata stores a Map between each group name and data used in the metadata table (specified in PR #2022)
+#[derive(Default)]
+pub struct Metadata {
+    groups: LinkedHashMap<Id, (usize, PathBuf)>,
+}
+
+impl Metadata {
+    /// Create an empty metadata table
+    pub fn new() -> Self {
+        Metadata {
+            groups: LinkedHashMap::new(),
+        }
+    }
+
+    /// Add a new entry to the metadata table
+    fn add_entry(&mut self, name: Id, line: usize, path: PathBuf) -> &mut Self {
```

Comment:
I don't see why this method should return a mutable reference to the struct
especially since it doesn't seem to be used.